lib/src/lifecycle_module.dart

```diff
@@ -725,25 +731,30 @@ abstract class LifecycleModule extends SimpleModule
 
       ShouldUnloadResult shouldUnloadResult = shouldUnload();
       if (!shouldUnloadResult.shouldUnload) {
+        _state = _previousState;
+        _previousState = null;
+        _transition = null;
         // reject with shouldUnload messages
         throw new ModuleUnloadCanceledException(
             shouldUnloadResult.messagesAsString());
       }
-
       _willUnloadController.add(this);
-      await Future.wait(_childModules.map((child) => child.unload()));
+      for (var child in _childModules.toList()) {
+        await child.unload();
+      }
+      _childModules.clear();
       await onUnload();
       await _disposableProxy.dispose();
       if (_state == LifecycleState.unloading) {
         _state = LifecycleState.unloaded;
+        _previousState = null;
         _transition = null;
       }
       _didUnloadController.add(this);
+      await _didUnloadController.close();
```

Comment:
Removing this from the try block introduces a possible leak in the event of an exception. The behavior of the module may be undefined in this case, but we should be sure to cleanup the stream controller to prevent leaks.